sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala

```diff
@@ -45,6 +47,41 @@ object CharVarcharUtils {
     })
   }
 
+  /**
+   * Validate the given schema to fail if it contains char or varchar types
+   *
+   * @param schema the given struct type
+   * @return
+   */
+  def failWithCharLikeTypes(schema: StructType): StructType = {
+    if (schema.exists(f => hasCharVarchar(f.dataType) ||
+      f.metadata.contains(CHAR_VARCHAR_TYPE_STRING_METADATA_KEY))) {
```

Comment:
when do we need to check the metadata?

Reply:
We can assume that users won't set the very internal `__CHAR_VARCHAR_TYPE_STRING` metadata key, and only check with the explicit char/varchar type.